apps/HelloiOS/HelloiOS/HalideView.mm

```diff
@@ -0,0 +1,364 @@
+    
+    int32_t iteration;
+    double frameElapsedEstimate;
+}
+
+
+#if HAS_METAL_SDK
++ (Class)layerClass
+{
+    return [CAMetalLayer class];
+}
+#endif  // HAS_METAL_SDK
+
+- (void)initCommon
+{
+    self.opaque          = YES;
+    self.backgroundColor = nil;
+
+#if HAS_METAL_SDK
+    _metalLayer = (CAMetalLayer *)self.layer;
+    _metalLayer.delegate = self;
+
+    _device = MTLCreateSystemDefaultDevice();
+    _commandQueue = [_device newCommandQueue];
+    
+    _metalLayer.device      = _device;
+    _metalLayer.pixelFormat = MTLPixelFormatBGRA8Unorm;
+    
+    _metalLayer.framebufferOnly = NO;
+#endif  // HAS_METAL_SDK
+}
+
+- (void) resetFrameTime
+{
+    iteration = 0;
+    frameElapsedEstimate = -1;
+}
+
+- (void) updateFrameTime: (double) elapsed using_metal: (bool)using_metal
+{
+    // Smooth elapsed using an IIR
+    if (frameElapsedEstimate == -1) {
+        frameElapsedEstimate = elapsed;
+    } else {
+        frameElapsedEstimate = (frameElapsedEstimate * 31 + elapsed) / 32.0;
+    }
+    if ((iteration % 30) == 0) {
+        dispatch_async(dispatch_get_main_queue(), ^(void) {
+            [self updateLogWith: frameElapsedEstimate using_metal: using_metal];
+        });
+    }
+    iteration += 1;
+}
+
+- (void)initBufsWithWidth: (int)w height: (int)h using_metal: (bool) using_metal
+{
+    // Make a pair of buffers to represent the current state
+    if (using_metal) {
+        buf1 = Buffer<float>::make_interleaved(w, h, 3);
+        buf2 = Buffer<float>::make_interleaved(w, h, 3);
+    } else {
+        buf1 = Buffer<float>(w, h, 3);
+        buf2 = Buffer<float>(w, h, 3);
+    }
+
+    // We really only need to pad this for the use_metal case,
+    // but it doesn't really hurt to always do it.
+    const int c = 4;
+    const int pad_pixels = (64 / sizeof(int32_t));
+    const int row_stride = (w + pad_pixels - 1) & ~(pad_pixels - 1);
+    const halide_dimension_t pixelBufShape[] = {
+        {0, w, c},
+        {0, h, c * row_stride},
+        {0, c, 1}
+    };
+
+    // This allows us to make a Buffer with an arbitrary shape
+    // and memory managed by Buffer itself
+    pixel_buf = Buffer<uint8_t>(nullptr, 3, pixelBufShape);
+    pixel_buf.allocate();
+}
+
+- (void)didMoveToWindow
+{
+    self.contentScaleFactor = self.window.screen.nativeScale;
+}
+
+- (id)initWithFrame:(CGRect)frame
+{
+    self = [super initWithFrame:frame];
+    if(self)
+    {
+        [self initCommon];
+    }
+    return self;
+}
+
+- (instancetype)initWithCoder:(NSCoder *)coder
+{
+    self = [super initWithCoder:coder];
+    
+    if(self)
+    {
+        [self initCommon];
+    }
+    return self;
+}
+
+- (void)updateLogWith: (double) elapsedTime using_metal: (bool) using_metal
+{
+#if HAS_METAL_SDK
+    const char* mode = using_metal ? "(Metal; Double-tap for CPU)" : "(CPU; Double-tap for Metal)";
+#else
+    const char* mode = "(CPU; Metal not available)";
+#endif
+    char log_text[2048];
+    snprintf(log_text, sizeof(log_text),
```

Comment:
It's probably more canonical to use NSString directly for constructing these messages, rather than C stdio methods. Cf. https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/Strings/Articles/FormatStrings.html